---
title: "Filtering"
output: html_notebook
---
### initialize


```{r}
library(here)
rm(list=ls())
working_dir <- paste(here(), "/filtering", sep="")
setwd(working_dir)
```

```{r "setup", include=FALSE}
require("knitr")
opts_knit$set(root.dir = working_dir)
```


```{r Function and vcftools initialization}
vcf_file <- list.files(pattern="*.vcf")[1]

snp_count <- function(vcf) {system(paste("mawk '!/#/'", as.character(vcf), "| wc -l", sep=" "), intern=T)}
snp_diff <- function(vcf1, vcf2) {
  message(
    snp_count(vcf2),
    " SNPs kept. ",
    as.numeric(snp_count(vcf1)) - as.numeric(snp_count(vcf2)),
    "(",
    round(100 - (as.numeric(snp_count(vcf2)) / as.numeric(snp_count(vcf1))) * 100, 2),
    "%) filtered. Cummulative filtered: ",
    as.numeric(snp_count(vcf_file)) - as.numeric(snp_count(vcf2)),
    "(",
    round(100 - (as.numeric(snp_count(vcf2)) / as.numeric(snp_count(vcf_file))) * 100, 2),
    "%)"
  )
}

#determine vcftools version
version <-system('vcftools', intern=T)[2]

#Path to vcftools v0.1.12 or higher, edit if warned below
vcf_PATH <- "export PATH=<PATH_TO_VCFTOOLS_0.1.12_or_later>:$PATH;"
#example
#vcf_PATH <- " export PATH=/home/evan/miniconda2/bin/:$PATH;"

#
if(grepl("0.1.11", version)) {
     assign("vcftools_path_export",vcf_PATH)
     message("Please update vcftools to v0.1.12 or later, or edit <PATH> above")
   } else {
     assign("vcftools_path_export", "")
     message("using VCFtools v0.1.12 or higher")
   }
```


```{r}
system("curl -L -O https://github.com/jpuritz/dDocent/raw/master/scripts/ErrorCount.sh")
system("chmod +x ErrorCount.sh")
```


```{r}
message(paste(snp_count(vcf_file), "Starting SNPs", sep=" "))
system(paste("./ErrorCount.sh", vcf_file, sep=" "))
```


### Filter by quality, minimum allele count, depth.

```{r}
#--mac filter skipped. Rare minor alleles expected in pooled data
system(paste(vcftools_path_export,"vcftools --vcf", vcf_file, "--minQ 30 --minDP 4 --recode --recode-INFO-all --out bare"))
snp_diff(vcf_file, "bare.recode.vcf")

system(paste("./ErrorCount.sh", "bare.recode.vcf", sep=" "))
```

### 5-step dDcoent filters

```{r}

#Allele balance filter skipped


#Ratio of mapping qualities between reference and alternate alleles - filter skipped. MQ indicated mapping quality, which is highly influenced by uniqueness of mapped reads. In libraries of sequenced individuals, uniquely mapped reads can indicate sequencing error, however these can be more common in pooled libraries due to highly polymorphic sites being more common.

#Removes low quality, high depth sites
system(paste(vcftools_path_export, ' vcffilter -f "QUAL / DP > 0.25" bare.recode.vcf > Qual_DP.vcf'))
snp_diff("bare.recode.vcf", "Qual_DP.vcf")

system(paste("./ErrorCount.sh", "Qual_DP.vcf", sep=" "))
```

### Visualize 
```{r}
system(paste("cut -f8 Qual_DP.vcf | grep -oe ","DP=[0-9]*"," | sed -s 's/DP=//g' > F5.DEPTH"))
    
system("mawk '!/#/' Qual_DP.vcf | cut -f1,2,6 > F5.loci.qual")
dp1 <- system("mawk '{ sum += $1; n++ } END { if (n > 0) print sum / n; }' F5.DEPTH", intern=T)

dp2 <- system(paste('python -c "print int(', dp1,'+3*(', dp1,'**0.5))"', sep=""), intern=T)

system(paste("paste F5.loci.qual F5.DEPTH | mawk -v x=",dp2," '$4 > x' | mawk '$3 < 2 * $4' > F5.lowQDloci", sep=""))

system(paste(vcftools_path_export, ' vcftools --vcf Qual_DP.vcf --site-depth --exclude-positions F5.lowQDloci --out F6'))
snp_diff("Qual_DP.vcf", "F6.ldepth")

system("cut -f3 F6.ldepth > F5.site.depth")
system("mawk '!/D/' F5.site.depth | mawk -v x=4 '{print $1/x}' > meandepthpersite")
```

```{sh}
gnuplot << \EOF 
set terminal dumb size 120, 30
set autoscale
set xrange [-10:100] 
unset label
set title "Histogram of mean depth per site"
set ylabel "Number of Occurrences"
set xlabel "Mean Depth"
binwidth=2
bin(x,width)=width*floor(x/width) + binwidth/2.0
set xtics 5
plot 'meandepthpersite' using (bin($1,binwidth)):(1.0) smooth freq with boxes
pause -1
EOF
```

```{r}
system(paste(vcftools_path_export, " vcftools --vcf  Qual_DP.vcf --recode-INFO-all --out F7 --max-meanDP 100 --exclude-positions F5.lowQDloci --recode"))
snp_diff("Qual_DP.vcf", "F7.recode.vcf")

system(paste("./ErrorCount.sh", "F7.recode.vcf", sep=" "))
```


```{r}
system("vcfallelicprimitives F7.recode.vcf --keep-info --keep-geno > filtered.prim.vcf")
snp_diff("F7.recode.vcf", "filtered.prim.vcf")
snp_count("filtered.prim.vcf")

system(paste(vcftools_path_export, " vcftools --vcf filtered.prim.vcf --remove-indels --recode --recode-INFO-all --out SNP.filtered"))
snp_diff(" filtered.prim.vcf", "SNP.filtered.recode.vcf")

system("cp SNP.filtered.recode.vcf ../filtered_SNPS.vcf")
system("cp Qual_DP.vcf ../bare_filtered_SNPS.vcf")
```


